### ISSUE TITLE
clang-tidy misc-confusable-identifiers check does not diagnose confusability due to normalization

### ISSUE THREAD

**Issue by zygoloid** [llvm/llvm-project#193140]
[Testcase](https://godbolt.org/z/4Kzna9qo6):
```c++
int ü = 1;
int ü = 2;

int fool = 1;
int foo1 = 2;
```
GCC has an enabled-by-default warning for the first case. You probably can't see what's happening here: in the first line, the variable name is in NFC (a single code point: u with diaresis), whereas in the second line the variable name is in NFD (two code points: u and a combining diaresis). These are distinct identifiers, so that's not a redeclaration of the first identifier. Neither clang nor clang-tidy has any check for this.

clang-tidy's `misc-confusable-identifiers` check does catch the second case. Depending on your font, that one might be more obvious: the first declaration ends in an L, whereas the second ends in a 1. `misc-confusable-identifiers` should catch the first case too; I'm surprised it's not already caught there, given that [the skeleton algorithm](https://www.unicode.org/reports/tr39/tr39-3.html#Confusable_Detection) for confusable detection starts by converting the identifiers to NFD, so both identifiers should map to the same skeleton and hence trigger the warning.

**Comment by zygoloid** [comment]
@serge-sans-paille 

https://github.com/llvm/llvm-project/blob/83f8eee57d5a2e9215b506a5ddc82db980fb47d0/clang-tools-extra/clang-tidy/misc/ConfusableIdentifierCheck.cpp#L40-L46

I'm not sure whether this can lead to false positives or not, but it seems to be leading to false negatives.

**Comment by zygoloid** [comment]
For what it's worth, @cor3ntin [brought this up](https://reviews.llvm.org/D112916#3595459) on the original review. Does LLVM have Unicode normalization tables that we could use now?

**Comment by cor3ntin** [comment]
We still do not have access to such table - I should work on that but it's a non trivial amount of work to upstream.

(We need both a code generator to produce reasonably optimized table and the normalization algorithm - it's something I've done a couple of times, it's a doozy but also [not a terrible amount of code](https://github.com/cor3ntin/ext-unicode-db/blob/cpp_gen/src/cedilla/normalization.hpp))

We are also not conforming for the grammar of identifiers because of that - we are supposed to reject non-nfc identifiers.

**Comment by zygoloid** [comment]
Oh interesting, so there's a rejects-valid bug in clang here too. Was the NFC requirement added as a DR?

**Comment by cor3ntin** [comment]
> Oh interesting, so there's a rejects-valid bug in clang here too. Was the NFC requirement added as a DR?

Nah, part of https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2021/p1949r7.html (we should track that more clearly)
